keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -248,11 +458,833 @@ Package `kuberuntime`
+
+> PreStop Hooks will be sent to sidecars before containers are terminated. This will be useful in scenarios such as when your sidecar is a proxy so that it knows to no longer accept inbound requests but can continue to allow outbound ones until the the primary containers have shut down.
+
+In other words, the shutdown sequence proposed is:
+
+1. Run preStop hooks for sidecars
+1. Run preStop hooks for non-sidecars
+1. Send SIGTERM for non-sidecars
+1. Run preStop hooks for sidecars
+1. Send SIGTERM for sidecars
+
+For brevity, what happens when some step timeouts and needs to send a SIGKILL is
+omitted in the above sequence, as it is not relevant for this point.
+
+The concerns we see with this are that this changes the [current delivery
+guarantees][hook-delivery] from _at least once_ to _at least twice_ for _some_
+containers (for sidecar containers). Furthermore, before this patch preStop
+hooks were delivered only once most of the time. After this patch is delivered
+twice for most cases (using sidecars). The problem is not if this is idempotent
+or not (as it should be in any case), but chainging the most common case from
+delivering preStop hook once to twice.
+
+Another concern is that in the future a different container type might be added,
+and these semantics seems difficult to extend. For example, let's say a
+container type "OrderedPhases" with the semantics of the [explained
+alternative][phase-startup] is added in the future. When would the preStop hooks
+be executed now that there are several phases? At the beginning? If there are 5
+phases: 0, 1, 2, 3 and 4, how should the shutdown behaviour be? Run preStop
+hooks for containers marked in phase 0, then kill containers in phase 4, then
+preStop hooks for containers in phase 1, then kill containers in phase 3, etc.?
+Or only the preStop hooks for container 0 are called? Why?
+
+It seems confusing, there doesn't seem to be a clear answer on those cases nor
+what the use case would be for such semantics.
+
+[hook-delivery]: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-delivery-guarantees
+[phase-startup]: #pre-defined-phases-for-container-startupshutdown-or-arbitrary-numbers-for-ordering
+
+##### Alternative 1: add a TerminationHook
+
+Add to containers a `TerminationHook` field. It will accept the same values as
+preStop hooks and will be called for any container that defines it when the pod
+is switching to a terminating state. These clear semantics seem easy to extend in
+the future.
+
+Then, instead of running preStop hooks twice for sidecar containers as it is now
+proposed, preStop hooks are run only once: just before stopping the
+corresponding containers.
+
+The shutdown sequence steps in this case would be:
+1. Run TerminationHook for _any_ container that defines it (sidecar or not)
+1. Run preStop hooks for non-sidecars
+1. Send SIGTERM for non-sidecars
+1. Run preStop hooks for sidecars
+1. Send SIGTERM for sidecars
```

Comment:
@howardjohn Another question, if it is not too much trouble: will this alternative work for you guys?

To make it short: the current KEP proposes to run preStop hooks for sidecars as the first step in the shutdown sequence, to let service mesh drain traffic. This PR proposes (the reasons are explained ^) to modify that behavior and have a `TerminationHook` that is run first. It will be the same, just under a new field called `TerminationHook`.

IIUC, this is needed for some service mesh (I'm working with a client with a in-house service mesh that needs this, for example) and I think Istio needs it, but wanted to confirm.

Does Istio needs this and would it work for you guys? 


PS: Linkerd, for example, doesn't need this as they use kubernetes services and pods in terminating state are removed from the `endpoints` objects (except for services with `externalTrafficPolicy: Local`, but that is addressed in another KEP), so that alone drains the service mesh connections. Not 100% sure in which case Istio currently is.

Reply:
I don't think we have a need to run any prestop hooks. I am not sure why we would start draining traffic. Isn't that what the KEP is fixing? The application shuts down, at which point all connections are closed. Then we can just terminate the sidecar immediately.

We certainly need this draining *today* but I don't think we do if this is implemented.

Reply:
Don't you need to de-register the pod from service discovery or something, when pod is about to terminate, so the app can just finish in-flight request and stop, without receiving new incoming requests via the service mesh?

On shutdown just having:
1. SIGTERM non-sidecars containers (using preStop hooks and all as expected)
1. SIGTERM sidecars containers (using preStop hooks and all as expected)

will work for istio?

Reply:
Without Istio when pod termination is triggered:
* Pod starts to terminate
* Pod is removed from Endpoint
* Endpoint update is sent to kube-proxy, stops sending traffic to this pod
* Pod fully shuts down

Hopefully in that order, but not guaranteed!! There is no guarantee of zero downtime here. See https://blog.sebastian-daschner.com/entries/zero-downtime-updates-kubernetes for more detailed explaination.

With Istio when pod termination is triggered:
* Pod starts to terminate
* Pod is removed from Endpoint
* Endpoint update is sent to istio-proxy, stops sending traffic to this pod
* Pod fully shuts down

You'll notice this is really the same. 

A valid solution to making the race conditions mentioned in the blog is adding a preStop hook - it doesn't have to be the sidecars preStop hook though. Any of:
* Telling users to add preStop hook if they want this behavior (this doesn't violate the transparency goal of service mesh, as the same issue is present without service mesh)
* Add the preStop hook to our sidecar
* Add the preStop hook to the users container (we control injection, we can modify whatever we want).

Reply:
Thanks for the detailed answer!

@Joseph-Irving do you remember why that part (running preStop hook twice) was added to the KEP?

Reply:
This was suggested by someone from istio, convo starts here https://github.com/kubernetes/community/pull/2148#issuecomment-422429820
```
the desired behavior would be for the sidecar to receive an event that the pod is shutting down and immediately put Envoy into lameduck mode, where it will reject inbound requests.
```
The idea being that you send a pre-stop to envoy so it would be able to reject incoming requests while the app is shutting down
If this is no longer required I'd be more than happy to remove it from the design.

Reply:
@nmittler

Reply:
I'm not sure it's still needed, given the current state of this proposal. If we can guarantee that the sidecar will be up for the lifetime of the application, that's probably good enough. If we need graceful shutdown/lameducking, we can probably handle that via the application container hooks.

Reply:
What do you mean the application container hooks? The istio container or app containers?

As you can see in the diff, istio pre-stop hook will only be called after main app terminated if we don't add the first step. So, it won't be possible to do something *before* if we remove it.

If you think it is not needed, though, we can remove it and add it in alpha stage if something comes up. Does it sound reasonable? :)

Reply:
I was referring to handling SIGTERM in the application container (not the sidecar) to initiate lameduck of the sidecar. That seems like a reasonable solution and will help focus this discussion just on start-up ordering.

Reply:
@nmittler The discussion is on startup and shutdown ordering (this KEP changes both), so this seems relevant, IMHO. Also, changing this has consequences on other related problems (like [this one][link]).

How would that script you have in mind look like? I mean, is it a istio binary that applications using istio need to bundle on their own container images, then? That seems like too much coupling, IMHO.

For other in house service mesh, I also see the need for the `TerminationHook`, for the very same reasons stated in the KEP.

If the alternative you have in mind is making the app have some service-mesh specific hooks and binaries, I think it seems more reasonable to me to have the `TerminationHook`. That way, the service mesh can run it's binaries (no coupling, binaries might change on different version of the service mesh), is possible to auto-inject containers (you can't auto-inject if you need to add a binary to the container image of the application), etc.

What do you think? Am I missing something?


[link]: https://github.com/kubernetes/enhancements/pull/1913/files#diff-c6212b56619f2b462935ad5f631d772fR691

Reply:
@rata to be clear, I'm not saying that `TerminationHook` wouldn't be useful, I suspect it would be. It just isn't clear that it would be absolutely necessary for Istio, assuming the sidecar is guaranteed to be up and ready for the duration of the application container.

Effectively, I'm echoing @howardjohn's ["valid solution"](https://github.com/kubernetes/enhancements/pull/1913#discussion_r463286867), which just uses `preStop` hooks. Specifically, the option where we inject a `preStop` hook into the application container, which effectively puts the sidecar in lameduck mode. This would basically just be a curl command.

If `TerminationHook` were available, however, it would be cleaner WRT injection since we could handle lameducking in the sidecar's `TerminationHook` rather than mucking about with the application container.

Reply:
@nmittler Cool, thanks a lot! A curl command seems like a trade-off worth considering. Thanks again for the input :-)

For other reviewers, here goes a tl; dr: The `TerminationHook` field is a proposal to remove calling two times the preStop hooks, due to the problems that create (see this very same section the comment is on for additional context). Given that implementation-wise is quite similar to what we already have in the code, I think keeping it can be simple too.

Just let us know if you think it will be simpler to remove it for now (either from an implementation POV or from a KEP design POV) so we settle on one or the other.

The main reason for this 2 years ago was istio, but there are today some other in-house service mesh apps I'm working with that must have this to work properly. But Istio might be fine without it to start with :)

Reply:
The concept of a Termination hook may be broadly useful so perhaps it makes sense to remove it from this KEP and develop it independently? It has no dependency on this proposal. It would also simplify this KEP if we removed the pre-stop behaviour and just focus on startup/shutdown ordering.

Reply:
+1 ... that's really the thrust of much of my discussion above. I think a separate KEP makes a lot of sense here.

Reply:
I'm fine with either of the choices (keep it or leave it in this KEP) as long as we move forward with this KEP :-D. I'd like to know what @derekwaynecarr or others from SIG-node think?

Removing this from the KEP effectively removes a step from the shutdown sequence and that changes how we might want to split the time in the shutdown sequence. For example, this affects other callouts like [this one][link]). So, if this needs to be added as a different KEP, it will change the time for the shutdown sequence for sidecar containers as described in this KEP. IMHO, we can remove it for now but if we want to add it later, we may want to add it to this KEP.

[link]: https://github.com/kubernetes/enhancements/pull/1913/files#diff-c6212b56619f2b462935ad5f631d772fR691